activesupport/lib/active_support/messages/rotator.rb

```diff
@@ -11,10 +11,16 @@ def initialize(*args, on_rotation: nil, **options)
         @on_rotation = on_rotation
       end
 
-      def rotate(*args, **options)
+      def rotate(*args, on_rotation: nil, **options)
+        @on_rotation = on_rotation if on_rotation
         fall_back_to build_rotation(*args, **options)
       end
 
+      def on_rotation(&on_rotation)
+        @on_rotation = on_rotation
+        self
+      end
```

Comment:
Elaborating on what I said in https://github.com/rails/rails/issues/54357#issuecomment-2616371900, the intended API for <code>MessageVerifier<strong>s</strong></code> was the `#on_rotation` method, not `rotate(on_rotation: ...)`.  Also, the intended API for `MessageVerifier` was `new(on_rotation: ...)`, not `rotate(on_rotation: ...)`.

The idea was that only one `on_rotation` callback is called (not one per verifier), so that callback was specified in a singular way.

Reply:
Right, but then the use case exposed in https://github.com/rails/rails/issues/54357 isn't super convenient, unless I'm missing something?

Reply:
The holistic approach would be to configure `Rails.application.message_verifiers.on_rotation(...)`, and then initialize `signed_id_verifier` from that, similar to [how it is done](https://github.com/rails/rails/blob/fc906ea0b9aef363d3443e631108d027bb850092/activerecord/lib/active_record/railtie.rb#L328-L334) for `generated_token_verifier`.

I think adding a `MessageVerifier#on_rotation` method is okay, but it may be a foot gun when instances are shared between classes.  I checked, and `signed_id_verifier` is _not_ shared, but if someone wrote something like `Post.shared_verifier.on_rotation(...)`, it would inadvertently change `Comment.shared_verifier`.

I wouldn't support `rotate(on_rotation: ...)`, though, for the reasons previously mentioned.